bin/elasticurl/main.cpp

```diff
@@ -0,0 +1,540 @@
+                fprintf(stdout, "Response Status: %d\n", responseCode);
+                app_ctx.response_code_written = true;
+            }
+
+            for (size_t i = 0; i < len; ++i)
+            {
+                fwrite(header[i].name.ptr, 1, header[i].name.len, stdout);
+                fprintf(stdout, ": ");
+                fwrite(header[i].value.ptr, 1, header[i].value.len, stdout);
+                fprintf(stdout, "\n");
+            }
+        }
+    };
+    requestOptions.onIncomingBody = [&](Http::HttpStream &, const ByteCursor &data) {
+        fwrite(data.ptr, 1, data.len, app_ctx.output);
+    };
+
+    request.SetMethod(ByteCursorFromCString(app_ctx.verb));
+    request.SetPath(app_ctx.uri.path_and_query);
+
+    Http::HttpHeader host_header;
+    host_header.name = ByteCursorFromCString("host");
+    host_header.value = app_ctx.uri.host_name;
+    request.AddHeader(host_header);
+
+    Http::HttpHeader user_agent_header;
+    user_agent_header.name = ByteCursorFromCString("user-agent");
+    user_agent_header.value = ByteCursorFromCString("elasticurl_cpp 1.0, Powered by the AWS Common Runtime.");
+    request.AddHeader(user_agent_header);
+
+    std::shared_ptr<Aws::Crt::Io::StdIOStreamInputStream> bodyStream =
+        MakeShared<Io::StdIOStreamInputStream>(allocator, input_body, allocator);
+    int64_t data_len = 0;
+    if (aws_input_stream_get_length(bodyStream->GetUnderlyingStream(), &data_len))
+    {
+        fprintf(stderr, "failed to get length of input stream.\n");
+        exit(1);
+    }
+    if (data_len > 0)
+    {
+        std::string content_length = std::to_string(data_len);
+        Http::HttpHeader content_length_header;
+        content_length_header.name = ByteCursorFromCString("content-length");
+        content_length_header.value = ByteCursorFromCString(content_length.c_str());
+        request.AddHeader(content_length_header);
+        request.SetBody(bodyStream);
+    }
+
+    for (size_t i = 0; i < app_ctx.header_line_count; ++i)
+    {
+        char *delimiter = (char *)memchr(app_ctx.header_lines[i], ':', strlen(app_ctx.header_lines[i]));
+
+        if (!delimiter)
+        {
+            fprintf(stderr, "invalid header line %s configured.", app_ctx.header_lines[i]);
+            exit(1);
+        }
+
+        Http::HttpHeader user_header;
+        user_header.name = ByteCursorFromArray((uint8_t *)app_ctx.header_lines[i], delimiter - app_ctx.header_lines[i]);
+        user_header.value = ByteCursorFromCString(delimiter + 1);
+        request.AddHeader(user_header);
+    }
+
+    auto stream = connection->NewClientStream(requestOptions);
+    stream->Activate();
+
+    semaphore.wait(semaphoreULock, [&]() { return streamCompleted; });
+
+    connection->Close();
+    semaphore.wait(semaphoreULock, [&]() { return connectionShutdown; });
+
+    /* TODO: Wait until the bootstrap finishing shutting down, we may need to break the API for create the Bootstrap, a
```

Comment:
Right now, the bootstrap on_shutdown_complete callback is always NULL, which may need API change, and it may be the follow up PR or I can try in this PR.